File: eval-gen/src/validator.ts
import * as crypto from 'crypto';
import {
  DraftedQuestion,
  Assertion,
  GeneratedEvalItem,
  ValidationResult,
  QuestionCategory,
  DEFAULT_CATEGORY_WEIGHTS,
} from './types';
import { computeGroundingConfidence } from './answer-grounder';
import { isNearDuplicatePrompt, normalizePrompt } from './dedupe';

/**
 * Deduplicate questions by checking for near-identical prompts.
 * Uses simple normalized string comparison (embedding-based dedup can be added later).
 */
function deduplicateQuestions(items: GeneratedEvalItem[]): {
  deduplicated: GeneratedEvalItem[];
  removedCount: number;
} {
  const seen = new Set<string>();
  const deduplicated: GeneratedEvalItem[] = [];
  let removedCount = 0;

  for (const item of items) {
    const normalized = normalizePrompt(item.prompt);

    // Check for exact or near-duplicate
    let isDuplicate = false;
    for (const existing of seen) {
      if (existing === normalized || isNearDuplicatePrompt(normalized, existing)) {
        isDuplicate = true;
        break;
      }
    }

    if (!isDuplicate) {
      seen.add(normalized);
      deduplicated.push(item);
    } else {
      removedCount++;
    }
  }

  return { deduplicated, removedCount };
}

/**
 * Check category balance against target weights
 */
function checkCategoryBalance(
  items: GeneratedEvalItem[],
): Record<QuestionCategory, number> {
  const counts: Record<string, number> = {};
  for (const item of items) {
    counts[item.category] = (counts[item.category] || 0) + 1;
  }

  const balance: Record<QuestionCategory, number> = {} as Record<QuestionCategory, number>;
  for (const cat of Object.keys(DEFAULT_CATEGORY_WEIGHTS) as QuestionCategory[]) {
    balance[cat] = counts[cat] || 0;
  }
  return balance;
}

/**
 * Compute coverage: fraction of source rows the eval set actually exercises.
 *
 * Counts every row referenced as either a primary `source_location` or as
 * evidence via `referenced_rows` (resolved from supporting_fact_ids). The
 * denominator is bounded by `min(totalRows, items*3)` since each question
 * realistically touches ~1 primary + ~2 supporting rows; this keeps the
 * metric achievable when the eval count is much smaller than the dataset.
 */
function computeCoverage(
  items: GeneratedEvalItem[],
  totalRows: number,
): number {
  if (totalRows <= 0) return 0;
  const referencedRows = new Set<string>();
  for (const item of items) {
    if (item.source_location) referencedRows.add(item.source_location);
    for (const r of item.referenced_rows ?? []) {
      if (r) referencedRows.add(r);
    }
  }
  const realisticTouchBudget = Math.max(items.length * 3, 50);
  const denom = Math.min(totalRows, realisticTouchBudget);
  return denom > 0 ? referencedRows.size / denom : 0;
}

/**
 * Generate a stable ID from prompt + source_location
 */
function generateItemId(prompt: string, sourceLocation: string): string {
  const hash = crypto.createHash('sha256')
    .update(`${prompt}|${sourceLocation}`)
    .digest('hex');
  return hash.slice(0, 12);
}

/**
 * Convert drafted questions + assertions into GeneratedEvalItems
 */
export function buildEvalItems(
  questions: DraftedQuestion[],
  assertionMap: Map<number, Assertion[]>,
): GeneratedEvalItem[] {
  return questions.map((q, i) => {
    const referencedRows = new Set<string>(q.referenced_rows ?? []);
    if (q.source_location) referencedRows.add(q.source_location);
    return {
      id: generateItemId(q.prompt, q.source_location),
      prompt: q.prompt,
      expected_answer: q.expected_answer,
      source_location: q.source_location,
      assertions: assertionMap.get(i) ?? [],
      category: q.category,
      difficulty: q.difficulty,
      supporting_facts: q.supporting_facts ?? [],
      grounding_confidence: computeGroundingConfidence(q),
      referenced_rows: Array.from(referencedRows),
    };
  });
}

/**
 * Validate a generated eval set: dedup, check balance, check coverage
 */
export function validateEvalSet(
  items: GeneratedEvalItem[],
  totalRows: number,
): { validated: GeneratedEvalItem[]; result: ValidationResult } {
  const issues: string[] = [];

  // Deduplicate
  const { deduplicated, removedCount } = deduplicateQuestions(items);
  if (removedCount > 0) {
    issues.push(`Removed ${removedCount} duplicate question(s)`);
  }

  // Category balance
  const categoryBalance = checkCategoryBalance(deduplicated);
  const totalItems = deduplicated.length;
  for (const [cat, weight] of Object.entries(DEFAULT_CATEGORY_WEIGHTS)) {
    const actual = (categoryBalance[cat as QuestionCategory] || 0) / Math.max(1, totalItems);
    const expected = weight;
    if (Math.abs(actual - expected) > 0.15) {
      issues.push(`Category "${cat}" is ${actual < expected ? 'under' : 'over'}-represented (${Math.round(actual * 100)}% vs ${Math.round(expected * 100)}% target)`);
    }
  }

  // Coverage
  const coverageScore = computeCoverage(deduplicated, totalRows);
  const COVERAGE_TARGET = 0.75;
  const CLI_COUNT_CAP = 50; // matches CLI clamp in src/index.ts

  // Compute the actual unique-row count once for richer messaging
  const referencedRows = new Set<string>();
  for (const item of deduplicated) {
    if (item.source_location) referencedRows.add(item.source_location);
    for (const r of item.referenced_rows ?? []) {
      if (r) referencedRows.add(r);
    }
  }
  const uniqueRowsReferenced = referencedRows.size;

  const realisticTouchBudget = deduplicated.length * 3;
  const realisticMaxCoverage = totalRows > 0
    ? Math.min(1.0, realisticTouchBudget / totalRows)
    : 0;
  const recommendedCountForTarget = totalRows > 0
    ? Math.ceil((totalRows * COVERAGE_TARGET) / 3)
    : 0;
  const datasetSampledNotExhaustive = totalRows > 0
    && recommendedCountForTarget > CLI_COUNT_CAP;

  const pct = (v: number) => Math.round(v * 100);

  if (totalRows > 0 && deduplicated.length > 0 && coverageScore < COVERAGE_TARGET) {
    if (realisticMaxCoverage >= COVERAGE_TARGET) {
      // Coverage is achievable with current count — LLM clustered questions
      issues.push(
        `Coverage ${pct(coverageScore)}% (${uniqueRowsReferenced}/${totalRows} rows) is below the ${pct(COVERAGE_TARGET)}% target. ` +
        `The current count (${deduplicated.length}) can reach the target — re-run to encourage broader row spread, or bump --count slightly.`
      );
    } else if (recommendedCountForTarget <= CLI_COUNT_CAP) {
      // Achievable by modestly increasing --count
      issues.push(
        `Coverage ${pct(coverageScore)}% (${uniqueRowsReferenced}/${totalRows} rows). ` +
        `For ≥${pct(COVERAGE_TARGET)}% coverage on this dataset, increase --count from ${deduplicated.length} to ~${recommendedCountForTarget}.`
      );
    } else {
      // Dataset too large for exhaustive coverage — reframe as representative sample
      issues.push(
        `Dataset is large (${totalRows} rows) — exhaustive coverage isn't practical for an eval set. ` +
        `This eval set tests ${uniqueRowsReferenced} representative rows (${pct(coverageScore)}%). ` +
        `For broader testing, generate multiple eval sets with focused --description values targeting different segments of your data ` +
        `(e.g., by category, time period, or status).`
      );
    }
  }

  // Check for low-confidence items
  const lowConfidence = deduplicated.filter(i => i.grounding_confidence === 'low').length;
  if (lowConfidence > totalItems * 0.2) {
    issues.push(`${lowConfidence} question(s) have low grounding confidence`);
  }

  const passed = issues.length === 0 || (removedCount === 0 && coverageScore >= 0.2);

  return {
    validated: deduplicated,
    result: {
      passed,
      totalItems: deduplicated.length,
      duplicatesRemoved: removedCount,
      categoryBalance,
      coverageScore,
      issues,
      uniqueRowsReferenced,
      totalRows,
      realisticMaxCoverage,
      recommendedCountForTarget,
      datasetSampledNotExhaustive,
    },
  };
}

File: eval-ui-winui3/src/EvalToolkit.EvalGen/Pipeline/Validator.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using EvalToolkit.Core;

namespace EvalToolkit.EvalGen.Pipeline;

/// <summary>
/// Port of <c>eval-gen/src/validator.ts</c>. Converts drafted questions +
/// assertions into <see cref="GeneratedEvalItem"/>s, deduplicates by
/// normalized prompt, and computes category-balance / coverage validation.
/// </summary>
public static class Validator
{
    private const double CoverageTarget = 0.75;
    private const int CliCountCap = 50;

    /// <summary>
    /// Build <see cref="GeneratedEvalItem"/>s from drafted questions and
    /// their assertion map. Mirrors TS <c>buildEvalItems</c>.
    /// </summary>
    public static IReadOnlyList<GeneratedEvalItem> BuildEvalItems(
        IReadOnlyList<DraftedQuestion> questions,
        IReadOnlyDictionary<int, IReadOnlyList<Assertion>> assertionMap)
    {
        var result = new List<GeneratedEvalItem>(questions.Count);
        for (int i = 0; i < questions.Count; i++)
        {
            var q = questions[i];
            var referencedRows = new HashSet<string>(StringComparer.Ordinal);
            if (q.ReferencedRows is not null)
            {
                foreach (var r in q.ReferencedRows) referencedRows.Add(r);
            }
            if (!string.IsNullOrEmpty(q.SourceLocation)) referencedRows.Add(q.SourceLocation);

            result.Add(new GeneratedEvalItem
            {
                Id = GenerateItemId(q.Prompt, q.SourceLocation),
                Prompt = q.Prompt,
                ExpectedAnswer = q.ExpectedAnswer,
                SourceLocation = q.SourceLocation,
                Assertions = assertionMap.TryGetValue(i, out var asserts) ? asserts : Array.Empty<Assertion>(),
                Category = q.Category,
                Difficulty = q.Difficulty,
                SupportingFacts = q.SupportingFacts.ToList(),
                GroundingConfidence = AnswerGrounder.ComputeGroundingConfidence(q),
                ReferencedRows = referencedRows.ToList(),
            });
        }
        return result;
    }

    /// <summary>Validate a generated eval set: dedup, check balance, check coverage.</summary>
    public static (IReadOnlyList<GeneratedEvalItem> Validated, ValidationResult Result) ValidateEvalSet(
        IReadOnlyList<GeneratedEvalItem> items,
        int totalRows)
    {
        var issues = new List<string>();

        var (deduplicated, removedCount) = DeduplicateQuestions(items);
        if (removedCount > 0) issues.Add($"Removed {removedCount} duplicate question(s)");

        var categoryBalance = CheckCategoryBalance(deduplicated);
        var totalItems = deduplicated.Count;

        foreach (var (cat, weight) in QuestionCategories.DefaultWeights)
        {
            var count = categoryBalance.TryGetValue(cat, out var c) ? c : 0;
            var actual = (double)count / Math.Max(1, totalItems);
            if (Math.Abs(actual - weight) > 0.15)
            {
                issues.Add($"Category \"{cat.ToWireString()}\" is {(actual < weight ? "under" : "over")}-represented ({(int)Math.Round(actual * 100, MidpointRounding.AwayFromZero)}% vs {(int)Math.Round(weight * 100, MidpointRounding.AwayFromZero)}% target)");
            }
        }

        var coverageScore = ComputeCoverage(deduplicated, totalRows);
        var referencedRows = new HashSet<string>(StringComparer.Ordinal);
        foreach (var item in deduplicated)
        {
            if (!string.IsNullOrEmpty(item.SourceLocation)) referencedRows.Add(item.SourceLocation);
            if (item.ReferencedRows is not null)
            {
                foreach (var r in item.ReferencedRows)
                {
                    if (!string.IsNullOrEmpty(r)) referencedRows.Add(r);
                }
            }
        }
        var uniqueRowsReferenced = referencedRows.Count;

        var realisticTouchBudget = deduplicated.Count * 3;
        var realisticMaxCoverage = totalRows > 0
            ? Math.Min(1.0, (double)realisticTouchBudget / totalRows)
            : 0.0;
        var recommendedCountForTarget = totalRows > 0
            ? (int)Math.Ceiling((totalRows * CoverageTarget) / 3.0)
            : 0;
        var datasetSampledNotExhaustive = totalRows > 0 && recommendedCountForTarget > CliCountCap;

        int Pct(double v) => (int)Math.Round(v * 100, MidpointRounding.AwayFromZero);

        if (totalRows > 0 && deduplicated.Count > 0 && coverageScore < CoverageTarget)
        {
            if (realisticMaxCoverage >= CoverageTarget)
            {
                issues.Add(
                    $"Coverage {Pct(coverageScore)}% ({uniqueRowsReferenced}/{totalRows} rows) is below the {Pct(CoverageTarget)}% target. " +
                    $"The current count ({deduplicated.Count}) can reach the target — re-run to encourage broader row spread, or bump --count slightly.");
            }
            else if (recommendedCountForTarget <= CliCountCap)
            {
                issues.Add(
                    $"Coverage {Pct(coverageScore)}% ({uniqueRowsReferenced}/{totalRows} rows). " +
                    $"For ≥{Pct(CoverageTarget)}% coverage on this dataset, increase --count from {deduplicated.Count} to ~{recommendedCountForTarget}.");
            }
            else
            {
                issues.Add(
                    $"Dataset is large ({totalRows} rows) — exhaustive coverage isn't practical for an eval set. " +
                    $"This eval set tests {uniqueRowsReferenced} representative rows ({Pct(coverageScore)}%). " +
                    "For broader testing, generate multiple eval sets with focused --description values targeting different segments of your data " +
                    "(e.g., by category, time period, or status).");
            }
        }

        var lowConfidence = deduplicated.Count(i => i.GroundingConfidence == GroundingConfidence.Low);
        if (lowConfidence > totalItems * 0.2)
        {
            issues.Add($"{lowConfidence} question(s) have low grounding confidence");
        }

        var passed = issues.Count == 0 || (removedCount == 0 && coverageScore >= 0.2);

        return (deduplicated, new ValidationResult
        {
            Passed = passed,
            TotalItems = deduplicated.Count,
            DuplicatesRemoved = removedCount,
            CategoryBalance = categoryBalance,
            CoverageScore = coverageScore,
            Issues = issues,
            UniqueRowsReferenced = uniqueRowsReferenced,
            TotalRows = totalRows,
            RealisticMaxCoverage = realisticMaxCoverage,
            RecommendedCountForTarget = recommendedCountForTarget,
            DatasetSampledNotExhaustive = datasetSampledNotExhaustive,
        });
    }

    private static (IReadOnlyList<GeneratedEvalItem> Deduplicated, int RemovedCount) DeduplicateQuestions(
        IReadOnlyList<GeneratedEvalItem> items)
    {
        var seen = new List<string>();
        var deduplicated = new List<GeneratedEvalItem>();
        int removedCount = 0;

        foreach (var item in items)
        {
            var normalized = Dedupe.NormalizePrompt(item.Prompt);
            bool isDuplicate = false;
            foreach (var existing in seen)
            {
                if (existing == normalized || Dedupe.IsNearDuplicatePrompt(normalized, existing))
                {
                    isDuplicate = true;
                    break;
                }
            }
            if (!isDuplicate)
            {
                seen.Add(normalized);
                deduplicated.Add(item);
            }
            else
            {
                removedCount++;
            }
        }

        return (deduplicated, removedCount);
    }

    private static Dictionary<QuestionCategory, int> CheckCategoryBalance(IReadOnlyList<GeneratedEvalItem> items)
    {
        var balance = new Dictionary<QuestionCategory, int>();
        foreach (var cat in QuestionCategories.DefaultWeights.Keys)
        {
            balance[cat] = 0;
        }
        foreach (var item in items)
        {
            balance.TryGetValue(item.Category, out var c);
            balance[item.Category] = c + 1;
        }
        return balance;
    }

    private static double ComputeCoverage(IReadOnlyList<GeneratedEvalItem> items, int totalRows)
    {
        if (totalRows <= 0) return 0;
        var referencedRows = new HashSet<string>(StringComparer.Ordinal);
        foreach (var item in items)
        {
            if (!string.IsNullOrEmpty(item.SourceLocation)) referencedRows.Add(item.SourceLocation);
            if (item.ReferencedRows is not null)
            {
                foreach (var r in item.ReferencedRows)
                {
                    if (!string.IsNullOrEmpty(r)) referencedRows.Add(r);
                }
            }
        }
        var realisticTouchBudget = Math.Max(items.Count * 3, 50);
        var denom = Math.Min(totalRows, realisticTouchBudget);
        return denom > 0 ? (double)referencedRows.Count / denom : 0;
    }

    private static string GenerateItemId(string prompt, string sourceLocation)
    {
        var input = Encoding.UTF8.GetBytes($"{prompt}|{sourceLocation}");
        var hash = SHA256.HashData(input);
        var sb = new StringBuilder(12);
        for (int i = 0; i < 6; i++)
        {
            sb.Append(hash[i].ToString("x2", CultureInfo.InvariantCulture));
        }
        return sb.ToString();
    }
}
